--- Day 5: If You Give A Seed A Fertilizer ---
You take the boat and find the gardener right where you were told he would be: managing a giant "garden" that looks more to you like a farm.

"A water source? Island Island is the water source!" You point out that Snow Island isn't receiving any water.

"Oh, we had to stop the water because we ran out of sand to filter it with! Can't make snow with dirty water. Don't worry, I'm sure we'll get more sand soon; we only turned off the water a few days... weeks... oh no." His face sinks into a look of horrified realization.

"I've been so busy making sure everyone here has food that I completely forgot to check why we stopped getting more sand! There's a ferry leaving soon that is headed over in that direction - it's much faster than your boat. Could you please go check it out?"

You barely have time to agree to this request when he brings up another. "While you wait for the ferry, maybe you can help us with our food production problem. The latest Island Island Almanac just arrived and we're having trouble making sense of it."

The almanac (your puzzle input) lists all of the seeds that need to be planted. It also lists what type of soil to use with each kind of seed, what type of fertilizer to use with each kind of soil, what type of water to use with each kind of fertilizer, and so on. Every type of seed, soil, fertilizer and so on is identified with a number, but numbers are reused by each category - that is, soil 123 and fertilizer 123 aren't necessarily related to each other.

For example:

seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
The almanac starts by listing which seeds need to be planted: seeds 79, 14, 55, and 13.

The rest of the almanac contains a list of maps which describe how to convert numbers from a source category into numbers in a destination category. That is, the section that starts with seed-to-soil map: describes how to convert a seed number (the source) to a soil number (the destination). This lets the gardener and his team know which soil to use with which seeds, which water to use with which fertilizer, and so on.

Rather than list every source number and its corresponding destination number one by one, the maps describe entire ranges of numbers that can be converted. Each line within a map contains three numbers: the destination range start, the source range start, and the range length.

Consider again the example seed-to-soil map:

50 98 2
52 50 48
The first line has a destination range start of 50, a source range start of 98, and a range length of 2. This line means that the source range starts at 98 and contains two values: 98 and 99. The destination range is the same length, but it starts at 50, so its two values are 50 and 51. With this information, you know that seed number 98 corresponds to soil number 50 and that seed number 99 corresponds to soil number 51.

The second line means that the source range starts at 50 and contains 48 values: 50, 51, ..., 96, 97. This corresponds to a destination range starting at 52 and also containing 48 values: 52, 53, ..., 98, 99. So, seed number 53 corresponds to soil number 55.

Any source numbers that aren't mapped correspond to the same destination number. So, seed number 10 corresponds to soil number 10.

So, the entire list of seed numbers and their corresponding soil numbers looks like this:

seed  soil
0     0
1     1
...   ...
48    48
49    49
50    52
51    53
...   ...
96    98
97    99
98    50
99    51
With this map, you can look up the soil number required for each initial seed number:

Seed number 79 corresponds to soil number 81.
Seed number 14 corresponds to soil number 14.
Seed number 55 corresponds to soil number 57.
Seed number 13 corresponds to soil number 13.
The gardener and his team want to get started as soon as possible, so they'd like to know the closest location that needs a seed. Using these maps, find the lowest location number that corresponds to any of the initial seeds. To do this, you'll need to convert each seed number through other categories until you can find its corresponding location number. In this example, the corresponding types are:

Seed 79, soil 81, fertilizer 81, water 81, light 74, temperature 78, humidity 78, location 82.
Seed 14, soil 14, fertilizer 53, water 49, light 42, temperature 42, humidity 43, location 43.
Seed 55, soil 57, fertilizer 57, water 53, light 46, temperature 82, humidity 82, location 86.
Seed 13, soil 13, fertilizer 52, water 41, light 34, temperature 34, humidity 35, location 35.
So, the lowest location number in this example is 35.

What is the lowest location number that corresponds to any of the initial seed numbers?

In [1]:
puzzle_input = """seeds: 3139431799 50198205 3647185634 110151761 2478641666 139825503 498892555 8913570 961540761 489996751 568452082 100080382 907727477 42158689 1617552130 312026427 342640189 97088268 2049289560 336766062

seed-to-soil map:
1615836342 1401909974 23067952
785532007 269485885 88937774
3019002892 2773729385 10470414
4202163101 2747292152 26437233
3183210415 4217634159 77333137
2847460091 3211730218 136699600
2455891790 3791729773 70553041
3260543552 2581343101 165949051
3840286095 2849853212 361877006
4228600334 2361239030 66366962
1594559581 1077839137 21276761
380069408 165017790 44262617
3598718222 1894384162 241567873
0 1424977926 190757551
1894384162 2810496375 39356837
424332025 606264721 196539291
3521487829 2221977524 77230393
742681934 69797365 36566707
1638904294 1615735477 139190145
1335949488 0 69797365
779248641 802804012 6283366
2638766896 4008940964 208693195
250963029 1142644585 70452661
1933740999 3470280789 321448984
190757551 209280407 60205478
1778094439 1099115898 43528687
2255189983 3348429818 121850971
1000500225 809087378 268751759
1269251984 1754925622 66697504
874469781 358423659 126030444
2526444831 2135952035 86025489
2439072067 3992121241 16819723
3426492603 3897126015 94995226
1405746853 1213097246 188812728
321415690 106364072 58653718
2984159691 3862282814 34843201
2377040954 2299207917 62031113
3029473306 2427605992 153737109
2612470320 2784199799 26296576
620871316 484454103 121810618

soil-to-fertilizer map:
4245401761 2352458099 28057201
2099789767 3998256334 14950546
3446056574 2749719529 135349925
890092371 1379309857 42097049
953714890 896502554 10335567
3115342240 2380515300 218129381
3333471621 3885671381 112584953
663999152 0 226093219
1873325002 727305635 169196919
2042521921 1328150912 51158945
3581406499 4034715214 260252082
500989478 564295961 163009674
4273458962 4013206880 21508334
3992733429 2099789767 164092891
4156826320 2263882658 88575441
3841658581 2598644681 151074848
1094703999 226093219 21270249
1708571521 399542480 164753481
964050457 247363468 130653542
1236290130 1421406906 171284482
0 1592691388 500989478
932189420 378017010 21525470
2114740313 2885069454 1000601927
1115974248 1207835030 120315882
1407574612 906838121 300996909

fertilizer-to-water map:
3217858280 3761663130 355893932
2319366035 2401839275 72374872
1962726423 909927230 105011330
2115307878 441322644 204058157
2095064202 1824085445 20243676
110580631 329763915 34129824
2573484127 2701101998 225220022
1780224111 2342656863 59182412
1717605398 1571533532 62618713
3589165621 3062909078 75538793
842280446 1871096488 471560375
409726333 0 243114563
397401582 760576019 12324751
0 667815878 92760141
3990305711 2926322020 5981819
251664023 1741105813 32320840
2072629125 645380801 22435077
92760141 363893739 17820490
2798704149 3138447871 156185660
3664704414 2573484127 127617871
1839406523 381714229 59608415
283984863 1844329121 26767367
1313840821 1176023584 27579684
2391740907 1790738543 33346902
3929650615 4234312200 60655096
3996287530 3557602002 204061128
1402088224 772900770 137026460
1911435456 2474214147 51290967
1539114684 1393042818 178490714
3573752212 4117557062 15413409
1341420505 1014938560 60667719
3792322285 4132970471 6723091
4200348658 4139693562 80082634
652840896 1203603268 189439550
310752230 243114563 86649352
2425087809 1075606279 100417305
1899014938 1773426653 12420518
2067737753 1785847171 4891372
3799045376 2932303839 130605239
2954889809 3294633531 262968471
4280431292 4219776196 14536004
144710455 1634152245 106953568

water-to-light map:
1071107509 759231097 26724064
1293599454 642189614 64567949
3147690498 1633749175 71376364
3487999223 4080968704 155844998
1700873635 2097781236 292450760
1146121952 1950303734 147477502
2864027062 1470573702 75076585
2507471274 3021419800 175807670
2939103647 922829268 123372939
2826742681 2551285626 37284381
719212681 1821763210 100196810
4236813702 4292227071 2740225
3643844221 3837685117 243283587
209103905 785955161 52302440
100332146 752700417 6530680
1478192138 3229160565 10342236
1358167403 3239502801 94242072
3219066862 706757563 45942854
3062476586 1358475140 29661777
819409491 2390231996 26561329
4195136152 3511184798 41677550
1488534374 1046202207 35444248
3887127808 3552862348 284822769
2188804057 1705125539 116637671
0 309410320 100332146
1117778238 1921960020 28343714
261406345 3333744873 74978759
2100705169 1545650287 88098888
3450098653 3473284228 37900570
2742067948 1273800407 84674733
2305441728 107380774 202029546
3265009716 1130086491 143713916
1993324395 0 107380774
1523978622 465294601 176895013
106862826 2588570007 102241079
2683278944 864040264 58789004
4239553927 4236813702 55413369
1097831573 1081646455 19946665
524293914 1101593120 28493371
336385104 2833510990 187908810
552787285 3197227470 31933095
845970820 2690811086 142699904
988670724 1388136917 82436785
1452409475 838257601 25782663
3092138363 409742466 55552135
584720380 2416793325 134492301
4171950577 3450098653 23185575

light-to-temperature map:
2906633798 3843376160 451591136
1332454428 1190958320 69004583
1837712164 0 353313230
494809338 353313230 376619264
871428602 729932494 461025826
1401459011 1754772241 373416033
3976747173 3375456648 91164221
3495346659 3466620869 376755291
0 1259962903 494809338
2608238358 2459541635 298395440
3907558614 2984992977 69188559
3872101950 2286963246 35456664
4067911394 2757937075 227055902
3358224934 2322419910 137121725
2286963246 3054181536 321275112
1774875044 2128188274 62837120

temperature-to-humidity map:
3966168141 3406025946 214996780
4181164921 3292223571 113802375
1493139015 1471031672 367564898
1423475871 1838596570 69663144
0 479293006 226560784
2500785470 2859072453 433151118
3197453551 2500785470 96923792
758446483 1237739489 233292183
991738666 0 278789291
3555740534 3884539689 410427607
3294377343 2597709262 261363191
226560784 705853790 531885699
1860703913 305742584 20602508
2933936588 3621022726 263516963
1881306421 278789291 26953293
1270527957 326345092 152947914

humidity-to-location map:
848612454 2250862530 61410922
910023376 3689675651 35197452
3724873103 3865027106 240221283
483883727 3324946924 364728727
0 1766978803 483883727
1957894300 561533 922927950
945220828 2590144784 734802140
2880822250 0 561533
3447014853 1489120553 277858250
2881383783 923489483 565631070
3965094386 3724873103 140154003
1680022968 2312273452 277871332"""

test_input = """seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4"""

In [2]:
import sys

def get_maps(list_of_maps_as_string):
    list_of_maps = {}
    
    for map_as_string in list_of_maps_as_string:
        name_of_map = map_as_string.split(':')[0]
        rows_of_map = map_as_string.split('\n')[1:]

        # create map as dict 
        map_as_list_of_lists = []

        # fill map with map info from input
        for row in rows_of_map:

            # get row info and convert to understandable variables
            row_info = row.split(' ')

            destination = int(row_info[0]) 
            source = int(row_info[1])
            range_length = int(row_info[2])

            # alter map
            map_as_list_of_lists.append([destination, source, range_length])

        list_of_maps[name_of_map] = map_as_list_of_lists

    return list_of_maps

def get_map_value(map_as_list_of_lists, input_int):

    # check every rule of map
    for rule_as_list in map_as_list_of_lists:

        # rule_as_list[0] = destination
        # rule_as_list[1] = source 
        # rule_as_list[2] = range_length 

        # check if rule applies
        if input_int >= rule_as_list[1] and input_int < (rule_as_list[1] + rule_as_list[2]):

            # calculate return value
            deviation = input_int - rule_as_list[1]

            return rule_as_list[0] + deviation

    # if no rule applies return input value
    return input_int

In [3]:
# solve puzzle

def solve_puzzle(puzzle_input):
    
    def get_seeds_as_list(seed_info_as_string):
        
        # convert seeds to list of ints
        seed_list_as_string = seed_info_as_string.split(': ')[1]
        seeds_as_list_of_strings = seed_list_as_string.split(' ')

        seed_list = [int(seed_as_string) for seed_as_string in seeds_as_list_of_strings]
        
        return seed_list
    
    # split input on emtpy lines
    input_parts = puzzle_input.split('\n\n')
    
    # get seeds as list of int
    seed_list = get_seeds_as_list(input_parts[0])
    
    # get dict containing maps as dicts
    list_of_maps = get_maps(input_parts[1:])
    
    # calculate location for each seed by going through all maps
    current_lowest_location = sys.maxsize
    
    for seed in seed_list:
        soil = get_map_value(list_of_maps["seed-to-soil map"], seed)
        fertilizer = get_map_value(list_of_maps["soil-to-fertilizer map"], soil)
        water = get_map_value(list_of_maps["fertilizer-to-water map"], fertilizer)
        light = get_map_value(list_of_maps["water-to-light map"], water)
        temperatur = get_map_value(list_of_maps["light-to-temperature map"], light)
        humidity = get_map_value(list_of_maps["temperature-to-humidity map"], temperatur)
        location = get_map_value(list_of_maps["humidity-to-location map"], humidity)
        
        if location < current_lowest_location:
            current_lowest_location = location
    
    # return smallest location number
    return current_lowest_location

solve_puzzle(puzzle_input)

462648396

The solution: 462648396 is correct :)

--- Part Two ---
Everyone will starve if you only plant such a small number of seeds. Re-reading the almanac, it looks like the seeds: line actually describes ranges of seed numbers.

The values on the initial seeds: line come in pairs. Within each pair, the first value is the start of the range and the second value is the length of the range. So, in the first line of the example above:

seeds: 79 14 55 13
This line describes two ranges of seed numbers to be planted in the garden. The first range starts with seed number 79 and contains 14 values: 79, 80, ..., 91, 92. The second range starts with seed number 55 and contains 13 values: 55, 56, ..., 66, 67.

Now, rather than considering four seed numbers, you need to consider a total of 27 seed numbers.

In the above example, the lowest location number can be obtained from seed number 82, which corresponds to soil 84, fertilizer 84, water 84, light 77, temperature 45, humidity 46, and location 46. So, the lowest location number is 46.

Consider all of the initial seed numbers listed in the ranges on the first line of the almanac. What is the lowest location number that corresponds to any of the initial seed numbers?

In [65]:
def solving_second_part_of_puzzle(puzzle_input):
    
    def get_seeds_as_list_of_tupels(seed_info_as_string):
        
        # convert seeds to list of tupels
        seed_list_as_string = seed_info_as_string.split(': ')[1]
        seeds_as_list_of_strings = seed_list_as_string.split(' ')

        seed_list = [int(seed_as_string) for seed_as_string in seeds_as_list_of_strings]
        
        list_of_tupels = []
        
        for index in range(0, len(seed_list), 2):
            list_of_tupels.append((seed_list[index], seed_list[index + 1]))
        
        return list_of_tupels
    
    def get_seed_ranges(convercion_map, seed_with_range, res):
        
        if len(seed_with_range) == 0:
            return res
        
        starting_seed = seed_with_range[0]
        range_value = seed_with_range[1]
        
        if range_value == 0:
            return res
        
        # keep track of next rule if none applies to first element
        start_of_next_rule = starting_seed + range_value
        
        # go through each rule
        for index, rule in enumerate(convercion_map):
            
            destination = rule[0] 
            source = rule[1] 
            range_length_rule = rule[2]
            
            # check if rule applies to first+ elements
            if starting_seed >= source and starting_seed < source + range_length_rule:
                
                # if yes, get seed range for first+ elements 
                
                # get new range
                start_of_new_range = destination + (starting_seed - source)
                length_of_new_range = None
                
                # rule not covering the whole input
                # rule not covering everything if starting_seed + range_value > source + range_length
                if (starting_seed + range_value - 1) > (source + range_length_rule - 1):
                    length_of_new_range = source + range_length_rule - starting_seed
                
                # or input not in need of full rule range (rule is too long)
                elif (starting_seed + range_value - 1) <= (source + range_length_rule - 1):
                    length_of_new_range = range_value
                    
                res.append((start_of_new_range, length_of_new_range))
                
                # repeat this algo for remaining elements
                remaining_seed_with_rang = (starting_seed + length_of_new_range, range_value - length_of_new_range)
                res = get_seed_ranges(convercion_map, remaining_seed_with_rang, res)
                
                return res
            
            # if no, check if rule is in range 
            elif source > starting_seed and source < starting_seed + range_value:
                
                # if distance to first element samaller previouse rule
                if source < start_of_next_rule:
                    # save first element of this rule
                    start_of_next_rule = source
        
        # if you reach this, first element is not covered by any rule
        # save range from first element to first element - 1 covered by saved rule (this might be the while range)
        range_of_converted_seeds = start_of_next_rule - starting_seed
        res.append((starting_seed, range_of_converted_seeds))
        
        # repeat algo with remaining elements
        remaining_seed_with_rang = (start_of_next_rule, range_value - range_of_converted_seeds)
        res = get_seed_ranges(convercion_map, remaining_seed_with_rang, res)
        
        return res
        
        
    # split input on emtpy lines
    input_parts = puzzle_input.split('\n\n')
    
    # get seed ranges
    seed_and_ranges_in_list = get_seeds_as_list_of_tupels(input_parts[0])
    
    # get dict containing maps as dicts
    list_of_maps = get_maps(input_parts[1:])
    
    # map seed range to destination range
    
    list_of_map_names = ["seed-to-soil map",
                         "soil-to-fertilizer map", 
                         "fertilizer-to-water map",
                         "water-to-light map",
                         "light-to-temperature map",
                         "temperature-to-humidity map", 
                         "humidity-to-location map"
                        ]

    input_ranges = seed_and_ranges_in_list

    for map_name in list_of_map_names:
    
        # for each in put range generate the output range based on map
        maped_inputs = []

        for seed_range in input_ranges:
            maped_inputs = maped_inputs + get_seed_ranges(list_of_maps[map_name], seed_range, [])
        
        input_ranges = maped_inputs
    
    return min([loc_range[0] for loc_range in input_ranges])
    
    
solving_second_part_of_puzzle(puzzle_input)

2520479

The solution: 2520479 is correct :)